File: app/routers/ats_pipeline.py
"""ATS Pipeline router — PIP-008.

Mounts every service function from PIP-001..PIP-006 as HTTP endpoints.
Route ordering: literal-segment routes before any /{...} dynamic segments
so FastAPI's first-match-wins dispatch does not shadow literal paths.
"""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Response

from app.auth.roles import Role, normalize_role
from app.core.settings import S
from app.models import (
    PipelineEntryCreateIn,
    PipelineEntryOut,
    PipelineRatingIn,
    PipelineStatusChangeIn,
    PipelineStatusConfigIn,
    PipelineStatusConfigOut,
    PlacementIn,
    PlacementOut,
)
from app.services.ats_pipeline import (
    _require_flag,
    add_to_pipeline,
    change_status,
    get_pipeline_entry,
    get_placement,
    get_status_config,
    list_pipeline_for_candidate,
    list_pipeline_for_job,
    record_placement,
    remove_from_pipeline,
    set_rating,
    set_status_config,
)
from app.services.sessions import require_ui_session

router = APIRouter(prefix="/ui/ats/pipeline", tags=["ats-pipeline"])
router_admin = APIRouter(prefix="/ui/admin/ats", tags=["ats-pipeline-admin"])


# ---------------------------------------------------------------------------
# Probe — declared first so it is never shadowed by /{...} routes
# ---------------------------------------------------------------------------

@router.get("/feature-status")
async def pipeline_feature_status() -> dict:
    """Return the ATS pipeline enabled flag. Does NOT call _require_flag()."""
    return {"enabled": S.ats_pipeline_enabled}


# ---------------------------------------------------------------------------
# Status config (PIP-002)
# ---------------------------------------------------------------------------

@router.get("/statuses", response_model=PipelineStatusConfigOut)
async def list_pipeline_statuses(
    ctx: dict = Depends(require_ui_session),
) -> PipelineStatusConfigOut:
    return get_status_config()


# ---------------------------------------------------------------------------
# Bulk-list by job / candidate (PIP-001) — literal second segments before /{id}
# ---------------------------------------------------------------------------

@router.get("/by-job/{job_order_id}")
async def list_by_job(
    job_order_id: str,
    cursor: Optional[str] = None,
    limit: int = 50,
    ctx: dict = Depends(require_ui_session),
) -> dict:
    return list_pipeline_for_job(ctx["user_sub"], job_order_id, cursor=cursor, limit=limit)


@router.get("/by-candidate/{candidate_id}")
async def list_by_candidate(
    candidate_id: str,
    cursor: Optional[str] = None,
    limit: int = 50,
    ctx: dict = Depends(require_ui_session),
) -> dict:
    return list_pipeline_for_candidate(ctx["user_sub"], candidate_id, cursor=cursor, limit=limit)


# ---------------------------------------------------------------------------
# Pipeline entry CRUD (PIP-001)
# ---------------------------------------------------------------------------

@router.post("/entries", response_model=PipelineEntryOut, status_code=201)
async def create_pipeline_entry(
    body: PipelineEntryCreateIn,
    ctx: dict = Depends(require_ui_session),
) -> PipelineEntryOut:
    return add_to_pipeline(ctx["user_sub"], body)


@router.delete("/by-job/{job_order_id}/candidate/{candidate_id}", status_code=204, response_class=Response)
async def delete_pipeline_entry(
    job_order_id: str,
    candidate_id: str,
    ctx: dict = Depends(require_ui_session),
) -> Response:
    remove_from_pipeline(ctx["user_sub"], job_order_id, candidate_id)
    return Response(status_code=204)


# ---------------------------------------------------------------------------
# Status change (PIP-003)
# ---------------------------------------------------------------------------

@router.patch(
    "/by-job/{job_order_id}/candidate/{candidate_id}/status",
    response_model=PipelineEntryOut,
)
async def patch_pipeline_status(
    job_order_id: str,
    candidate_id: str,
    body: PipelineStatusChangeIn,
    ctx: dict = Depends(require_ui_session),
) -> PipelineEntryOut:
    return change_status(
        ctx["user_sub"], job_order_id, candidate_id,
        body.new_status, note=body.note,
    )


# ---------------------------------------------------------------------------
# Rating (PIP-004)
# ---------------------------------------------------------------------------

@router.patch(
    "/by-job/{job_order_id}/candidate/{candidate_id}/rating",
    response_model=PipelineEntryOut,
)
async def patch_pipeline_rating(
    job_order_id: str,
    candidate_id: str,
    body: PipelineRatingIn,
    ctx: dict = Depends(require_ui_session),
) -> PipelineEntryOut:
    return set_rating(ctx["user_sub"], job_order_id, candidate_id, body)


# ---------------------------------------------------------------------------
# Placement (PIP-006)
# ---------------------------------------------------------------------------

@router.post(
    "/by-job/{job_order_id}/candidate/{candidate_id}/placement",
    response_model=PlacementOut,
    status_code=201,
)
async def create_placement(
    job_order_id: str,
    candidate_id: str,
    body: PlacementIn,
    ctx: dict = Depends(require_ui_session),
) -> PlacementOut:
    return record_placement(ctx["user_sub"], job_order_id, candidate_id, body)


@router.get(
    "/by-job/{job_order_id}/candidate/{candidate_id}/placement",
    response_model=PlacementOut,
)
async def fetch_placement(
    job_order_id: str,
    candidate_id: str,
    ctx: dict = Depends(require_ui_session),
) -> PlacementOut:
    return get_placement(ctx["user_sub"], job_order_id, candidate_id)


# ---------------------------------------------------------------------------
# Admin: status config write (PIP-002)
# ---------------------------------------------------------------------------

@router_admin.put("/pipeline/statuses", response_model=PipelineStatusConfigOut)
async def update_pipeline_statuses(
    body: PipelineStatusConfigIn,
    ctx: dict = Depends(require_ui_session),
) -> PipelineStatusConfigOut:
    if normalize_role(ctx.get("role")) not in {Role.ADMIN, Role.ROOT}:
        raise HTTPException(403, detail={"code": "admin_role_required"})
    return set_status_config(ctx["user_sub"], body)

File: frontend/src/api/endpoints/atsPipeline.ts
/**
 * ATS Recruiting Pipeline API client — PIP-001..PIP-010.
 *
 * Backend prefix: /ui/ats/pipeline  (admin: /ui/admin/ats/pipeline)
 * Feature gate:   GET /ui/ats/pipeline/feature-status (no auth required)
 * All other endpoints require cookie-auth session.
 *
 * With ATS_PIPELINE_ENABLED=false the server returns 503 {code: "ats_pipeline_disabled"}.
 * Callers should handle ApiError.status === 503 gracefully.
 */
import { api } from "@/api/client";

// ─── Inline TypeScript types mirroring app/models.py ─────────────────────────

/** Pipeline entry — M:N junction between a candidate and a job order. */
export interface PipelineEntry {
  pipeline_id: string;
  job_order_id: string;
  candidate_id: string;
  owner_sub: string;
  /** e.g. "100_no_contact", "400_submitted", "900_placed" */
  status: string;
  /** Numeric rank from status config, used for board ordering */
  status_rank: number;
  /** 0 = unrated; 1–5 stars */
  rating: number;
  created_at: number;
  updated_at: number;
}

export interface PipelineEntryCreateIn {
  job_order_id: string;
  candidate_id: string;
  /** Defaults to "100_no_contact" if omitted */
  status?: string;
}

export interface PipelineStatusChangeIn {
  new_status: string;
  note?: string | null;
}

export interface PipelineRatingIn {
  /** 1–5 star rating; 0 clears the rating (unrated) */
  rating: number;
}

/** A single status stage in the pipeline ladder. */
export interface PipelineStatusItem {
  status_key: string;
  label: string;
  rank: number;
  order: number;
  is_submitted: boolean;
  is_placed: boolean;
  is_terminal: boolean;
  color: string | null;
}

export interface PipelineStatusConfig {
  statuses: PipelineStatusItem[];
  updated_at: number | null;
  updated_by_sub: string | null;
}

/** Admin-write payload — replaces the full status ladder. */
export interface PipelineStatusConfigIn {
  statuses: Array<{
    status_key: string;
    label: string;
    rank: number;
    order?: number;
    is_submitted?: boolean;
    is_placed?: boolean;
    is_terminal?: boolean;
    color?: string | null;
  }>;
}

/** Placement record — created when a candidate is formally placed. */
export interface Placement {
  placement_id: string;
  job_order_id: string;
  candidate_id: string;
  owner_sub: string;
  /** Unix epoch seconds for candidate start date */
  start_date: number;
  /** Integer cents; 0 = pro-bono */
  fee_cents: number;
  /** Status key at time of placement (e.g. "900_placed") */
  status_at_placement: string;
  notes: string | null;
  created_at: number;
}

export interface PlacementIn {
  start_date: number;
  fee_cents: number;
  notes?: string | null;
}

export interface PipelineListResponse {
  entries: PipelineEntry[];
  cursor: string | null;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Check whether the pipeline feature is enabled.
 * This probe does NOT call _require_flag() so it always returns (no 503).
 * NOTE: when ats_pipeline_enabled=false the router itself is not mounted
 * so this endpoint may 404. Treat 404 the same as enabled=false.
 */
export const getPipelineFeatureStatus = () =>
  api.get<{ enabled: boolean }>("/ui/ats/pipeline/feature-status");

// ─── Status Config (PIP-002) ──────────────────────────────────────────────────

export const getPipelineStatuses = () =>
  api.get<PipelineStatusConfig>("/ui/ats/pipeline/statuses");

/** Admin only — replace deployment-wide status ladder. */
export const setPipelineStatuses = (body: PipelineStatusConfigIn) =>
  api.put<PipelineStatusConfig>("/ui/admin/ats/pipeline/statuses", body);

// ─── Bulk list (PIP-001) ─────────────────────────────────────────────────────

export const listPipelineByJob = (
  jobOrderId: string,
  opts?: { cursor?: string; limit?: number },
) =>
  api.get<PipelineListResponse>(`/ui/ats/pipeline/by-job/${jobOrderId}`, {
    ...(opts?.cursor ? { cursor: opts.cursor } : {}),
    ...(opts?.limit ? { limit: String(opts.limit) } : {}),
  });

export const listPipelineByCandidate = (
  candidateId: string,
  opts?: { cursor?: string; limit?: number },
) =>
  api.get<PipelineListResponse>(`/ui/ats/pipeline/by-candidate/${candidateId}`, {
    ...(opts?.cursor ? { cursor: opts.cursor } : {}),
    ...(opts?.limit ? { limit: String(opts.limit) } : {}),
  });

// ─── Pipeline entry CRUD (PIP-001) ───────────────────────────────────────────

export const addToPipeline = (body: PipelineEntryCreateIn) =>
  api.post<PipelineEntry>("/ui/ats/pipeline/entries", body);

export const removePipelineEntry = (jobOrderId: string, candidateId: string) =>
  api.del<undefined>(`/ui/ats/pipeline/by-job/${jobOrderId}/candidate/${candidateId}`);

// ─── Status change (PIP-003) ─────────────────────────────────────────────────

export const changePipelineStatus = (
  jobOrderId: string,
  candidateId: string,
  body: PipelineStatusChangeIn,
) =>
  api.patch<PipelineEntry>(
    `/ui/ats/pipeline/by-job/${jobOrderId}/candidate/${candidateId}/status`,
    body,
  );

// ─── Rating (PIP-004) ────────────────────────────────────────────────────────

export const setPipelineRating = (
  jobOrderId: string,
  candidateId: string,
  body: PipelineRatingIn,
) =>
  api.patch<PipelineEntry>(
    `/ui/ats/pipeline/by-job/${jobOrderId}/candidate/${candidateId}/rating`,
    body,
  );

// ─── Placement (PIP-006) ─────────────────────────────────────────────────────

export const createPlacement = (
  jobOrderId: string,
  candidateId: string,
  body: PlacementIn,
) =>
  api.post<Placement>(
    `/ui/ats/pipeline/by-job/${jobOrderId}/candidate/${candidateId}/placement`,
    body,
  );

export const getPlacement = (jobOrderId: string, candidateId: string) =>
  api.get<Placement>(
    `/ui/ats/pipeline/by-job/${jobOrderId}/candidate/${candidateId}/placement`,
  );
